File: 백준/Gold/1043. 거짓말/거짓말.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

def find(x):
    if x == parent[x]:
        return x

    parent[x] = find(parent[x])
    return parent[x]


def union(a, b):
    a = find(a)
    b = find(b)

    if a in truth and b in truth:
        return
    
    if a in truth:
        parent[b] = a
    elif b in truth:
        parent[a] = b
    else:
        res = min(a, b)
        parent[a] = res
        parent[b] = res


n, m = map(int, input().split())
truth = set(list(map(int, input().split()))[1:])
party = []
parent = [i for i in range(n+1)]

for _ in range(m):
    people = list(map(int, input().split()))
    party.append(people[1:])

    for i in range(1, people[0]):
        union(people[i], people[i+1])

cnt = 0

for p in party:
    for person in p:
        if find(parent[person]) in truth:
            break
    else:
        cnt += 1

print(cnt)

File: 백준/Gold/1043. 거짓말/거짓말.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.StringTokenizer;

public class Main {
	private static int N, M, cnt;
	private static int[] root;
	private static HashSet<Integer> truth = new HashSet<>();
	
	private static int find(int num) {
		if (num == root[num]) {
			return num;
		}
		return root[num] = find(root[num]);
	}
	
	private static void union(int a, int b) {
		int aRoot = find(a);
		int bRoot = find(b);
		
		if (truth.contains(aRoot)) {
			root[bRoot] = aRoot;
		} else if (truth.contains(bRoot)) {
			root[aRoot] = bRoot;
		} else {
			root[bRoot] = aRoot;
		}
	}
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer nums = new StringTokenizer(br.readLine());
		N = Integer.parseInt(nums.nextToken());
		M = Integer.parseInt(nums.nextToken());
		cnt = 0;
		root = new int[N+1];
		
		for (int i = 0; i <= N; i++) {
			root[i] = i;
		}
		
		nums = new StringTokenizer(br.readLine());
		int t = Integer.parseInt(nums.nextToken());
		
		for (int i = 0; i < t; i++) {
			truth.add(Integer.parseInt(nums.nextToken()));
		}
		
		int[] party = new int[M];
		
		for (int i = 0; i < M; i++) {
			nums = new StringTokenizer(br.readLine());
			int n = Integer.parseInt(nums.nextToken());
			int a = Integer.parseInt(nums.nextToken());
			party[i] = a;
			
			for (int j = 0; j < n-1; j++) {
				int b = Integer.parseInt(nums.nextToken());
				union(a, b);
				a = b;
			}
		}
		
		for (int p : party) {
			if (truth.contains(find(p))) {
				cnt++;
			}
		}
		
		System.out.println(M - cnt);
	}

}